File: ruby/strategy/DuckSimulator/client/duck.rb
require './behavior/fly_behavior'
require './behavior/quack_behavior'

class Duck

  def initialize(duck_name)
    @duck_name = duck_name
    @fly_behavior = FlyBehavior.new
    @quack_behavior = QuackBehavior.new
  end

  def display
    puts("I am a " + @duck_name)
  end
  
  def swim
    puts(@duck_name + " is swimming")
  end
  
  def perform_fly
    @fly_behavior.fly
  end

  def perform_quack
    @quack_behavior.quack
  end

end

File: javascript/strategy/DuckSimulator/client/Duck.js
import { FlyBehavior } from "../behavior/FlyBehavior.js";
import { QuackBehavior } from "../behavior/QuackBehavior.js";

export class Duck {

    // private vars
    #duck_name

    constructor(duck_name) {
        this.#duck_name = duck_name
        this.fly_behavior = new FlyBehavior()
        this.quack_behavior = new QuackBehavior()
    }

    display() {
        console.log("I am a " + this.#duck_name)
    }

    swim() {
        console.log(this.#duck_name + " is swimming")
    }

    performFly() {
        this.fly_behavior.fly()
    }

    performQuack() {
        this.quack_behavior.quack()
    }

    setFly(fb) {
        this.fly_behavior = fb
    }

    setQuack(qb) {
        this.quack_behavior = qb
    }

}